src/Raven.Server/Documents/Replication/ReplicationTopologyDiscoveryHandler.cs

```diff
@@ -0,0 +1,127 @@
+                }
+
+                var token = tcp.DocumentDatabase.DatabaseShutdown;
+                using (var topologyDiscoverer = new ReplicationTopologyDiscoverer(
+                    outgoingHandlers,
+                    header.AlreadyVisitedDbIds,
+                    token))
+                {
+                    //if true, all adjacent nodes are already visited, so no need to continue
+                    //(filtering is done in a ctor of ReplicationTopologyDiscoverer, using header.AlreadyVisitedDbIds)
+                    if (topologyDiscoverer.DiscovererCount == 0)
+                    {
+                        WriteDiscoveryResponse(tcp, context, new List<TopologyNode>(), TopologyDiscoveryResponse.ResponseType.Ok);
+                        return;
+                    }
+
+                    var discoveryTask = topologyDiscoverer.DiscoverTopologyAsync();
+
+                    List<TopologyNode> nodes;
+                    TopologyDiscoveryResponse.ResponseType responseType;
+
+                    //if true, then the discovery is taking too long..
+                    //TODO : make topology discovery timeout configurable
+                    if (Task.WaitAny(discoveryTask, Task.Delay(TimeSpan.FromSeconds(60), token)) != discoveryTask.Id)
+                    {
+                        responseType = TopologyDiscoveryResponse.ResponseType.Timeout;
+                        nodes = Enumerable.Empty<TopologyNode>().ToList();
+                    }
+                    else
+                    {
+                        responseType = TopologyDiscoveryResponse.ResponseType.Ok;
+                        nodes = discoveryTask.Result.ToList();
+
+                        var replicationHandlers = tcp.DocumentDatabase.DocumentReplicationLoader.Outgoing;
+                        foreach (var node in nodes)
+                        {
+                            var relevantReplicationHandler = replicationHandlers.FirstOrDefault(h =>
+                                                                        h.DestinationDbId == node.Node.DbId);
+
+                            //precaution, if this will be true I will be surprised..
+                            if (relevantReplicationHandler == null)
+                            {
+                                throw new InvalidDataException($@"Couldn't find outgoing replication handler for DbId = {node.Node.DbId}. 
+                                                                  This is not supposed to happen and is likely a bug.");
+                            }
+
+                            node.SpecifiedCollections = relevantReplicationHandler.Destination.SpecifiedCollections;
+                            node.Disabled = relevantReplicationHandler.Destination.Disabled;
+                            node.IgnoredClient = relevantReplicationHandler.Destination.IgnoredClient;
+                            node.Node.ApiKey = relevantReplicationHandler.Destination.ApiKey;
```

Comment:
You do NOT send that kind of information over the network. Because just the fact that I let you replicate to me does NOT meant that you get to see the ApiKeys for my entire network.